### PR TITLE
Retry attach container failures

### LINKED ISSUE #5615: Add retries on CreateFile path
https://portal.microsofticm.com/imp/v3/incidents/details/232707912/home

The issue is that our CreateFile path has no retry logic, so if fetch() hits any issues (including call falling on the floor), the app gets failure. Currently Teams did retry logic on their side (but not OWA or OneNote), but that's wrong place to have it, it has to be in Container or driver.

### PR DESCRIPTION
**Pull request by jatgarg**
Fixes: https://github.com/microsoft/FluidFramework/issues/5615
This change enables to retry container attach on failures. This will hosts/apps not to retry at their end.

### CODE REVIEW

File: packages/loader/container-loader/src/container.ts [hunk 1]
```diff
@@ -753,6 +758,40 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
     }
 
     public async attach(request: IRequest): Promise<void> {
+        let success = false;
+        let numRetries = 0;
+        let retryAfter = 1;
+        let id: string | undefined;
+        do {
+            try {
+                await this.attachCore(request);
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/container.ts:767]: i don't think this is safe. we should be more specific about what we retry. probably only need to retry to the network call, and i think we usually do that in the driver

Comment (anthony-murphy) [reply packages/loader/container-loader/src/container.ts:767]: like i don;t think we can fire the attaching event multiple time, and we def can;t fire attaching after attached which is possible with the generalized retry here

**Comment by anthony-murphy** [comment]
please add tests

File: packages/loader/container-loader/src/container.ts (hunk 1 above)
Comment (jatgarg) [reply packages/loader/container-loader/src/container.ts:767]: But there is only 1 network call in the attach flow and if that succeeds then we don't expect anything else to fail. So if attach network call succeeds and anything else fails, then I don't think retrying would do much.

Comment (jatgarg) [reply packages/loader/container-loader/src/container.ts:767]: Added here as it is central place for attach rather than in all drivers.

Comment (jatgarg) [reply packages/loader/container-loader/src/container.ts:767]: We already take care of events firing. Like we only fire attaching if the container is detached in attach core.
I think I will add check to fire attached only from attaching.
If already attached, then we don't execute attach core as we just return.

Comment (anthony-murphy) [reply packages/loader/container-loader/src/container.ts:767]: just retry the network call then. also, needs tests regardless

Comment (jatgarg) [reply packages/loader/container-loader/src/container.ts]: Added tests and retried only the createContainer call in attach. Actually deduped some of the common code in the package.

**Comment by anthony-murphy** [comment]
this looks great! could you also add a test for the specific scenarion you are changing, take a look at packages\test\test-end-to-end-tests\src\test\sharedStringLoading.spec.ts for an example of how that could be done

**Review by anthony-murphy — APPROVED** [review]

**Pull request merged**